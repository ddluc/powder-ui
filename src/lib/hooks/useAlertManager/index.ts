import React from 'react';
import { AlertData } from '../../types';

export const useAlertManager = () => {
  const [alerts, setAlerts] = React.useState<AlertData[]>([]);

  const getAlertIndex = (id: string): number | null => {
    // Don't add item if it already exists
    const index = alerts.findIndex((val) => val.id === id);
    if (index >= 0) return index;
    return null;
  };

  const addAlert = (alert: AlertData): void => {
    const index = getAlertIndex(alert.id);
    // Don't add item if it exists in the hook state
    if (index != null) return;
    const result = [{ ...alert, duration: alert.duration } as AlertData, ...alerts];
    setAlerts(result);
  };

  const removeAlert = (id: string): void => {
    // Don't attempt to remove item if it doesn't exist in the hook state
    const index = getAlertIndex(id);
    if (index == null) return;
    const result = [...alerts.slice(0, index), ...alerts.slice(index + 1)];
    setAlerts(result);
  };

  return {
    addAlert,
    removeAlert,
    alerts
  };
};
