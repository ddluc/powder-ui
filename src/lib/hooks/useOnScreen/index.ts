import React, { useRef, useEffect, useState } from 'react';
import { debounce } from '../../util';

/**
 * Check if an element is in viewport
 * @param {number} offset - Number of pixels up to the observable element from the top
 * @param {number} throttleMilliseconds - Throttle observable listener, in ms
 */
export const useOnScreen = <Element extends HTMLElement>(
  offset = 0
): [boolean, React.RefObject<Element>] => {
  const [isVisible, setIsVisible] = useState(false);
  const currentElement = useRef<Element>();

  const onScroll = debounce(() => {
    if (!currentElement.current) {
      setIsVisible(false);
      return;
    }
    const { top } = currentElement.current.getBoundingClientRect();
    setIsVisible(top + offset >= 0 && top - offset <= window.innerHeight);
  });

  React.useEffect(() => {
    onScroll();
  }, []);

  useEffect(() => {
    document.addEventListener('scroll', onScroll, true);
    return () => document.removeEventListener('scroll', onScroll, true);
  });

  return [isVisible, currentElement];
};
