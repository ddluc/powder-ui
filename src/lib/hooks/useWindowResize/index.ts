import React from 'react';

export interface WindowSize {
  width: number | undefined;
  height: number | undefined;
}

export const useWindowSize = (): WindowSize => {
  const [windowSize, setWindowSize] = React.useState<WindowSize>({
    width: undefined,
    height: undefined
  });

  const handleResize = React.useCallback(() => {
    // Set window width/height to state
    setWindowSize({
      width: window.innerWidth,
      height: window.innerHeight
    });
  }, []);

  React.useEffect(() => {
    // Add event listener
    window.addEventListener('resize', handleResize);
    // Call handler right away so state gets updated with initial window size
    handleResize();
    // Remove event listener on cleanup
    return () => window.removeEventListener('resize', handleResize);
  }, []); // Empty array ensures that effect is only run on mount
  return windowSize;
};
