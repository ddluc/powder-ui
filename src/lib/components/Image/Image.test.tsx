import React from 'react';
import { render, cleanup } from '@testing-library/react';
import { Image, ImageProps } from './index';
import withTheme from '../../../../test/withTheme';

import * as hooks from '../../hooks/useOnScreen';

const args: ImageProps = {
  src: {
    main: 'https://piste-ui.s3.us-west-2.amazonaws.com/ski.jpg'
  },
  width: 400,
  height: 400,
  fit: 'cover',
  xpos: 50,
  ypos: 50
};

describe('Image', () => {
  jest.spyOn(hooks, 'useOnScreen').mockImplementation(() => [true, React.useRef()]);

  afterEach(cleanup);
  afterEach(() => jest.clearAllMocks());

  it('should render with defined props', () => {
    const result = render(<Image {...args} />);
    const component = result.container;
    expect(component).toMatchSnapshot();
  });

  it('should render with preview', () => {
    const src = {
      main: 'https://piste-ui.s3.us-west-2.amazonaws.com/ski.jpg',
      preview: 'https://piste-ui.s3.us-west-2.amazonaws.com/ski--thumbnail.jpg'
    };
    const result = render(<Image {...args} src={src} />);
    const component = result.container;
    expect(component).toMatchSnapshot();
  });

  it('should render as skeleton', () => {
    const result = withTheme(<Image skeleton height={400} width={400} />);
    const component = result.container;
    expect(component).toMatchSnapshot();
  });
});
