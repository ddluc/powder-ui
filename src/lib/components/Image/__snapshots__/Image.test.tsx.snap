// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image should render as skeleton 1`] = `
.c0 {
  width: 400px;
  height: 400px;
}

.c1 {
  height: 100%;
  width: 100%;
  background-color: #f9f9f9;
  border-radius: 3px;
  position: relative;
  padding: 0;
  box-shadow: none;
  pointer-events: none;
}

.c1:hover,
.c1:focus,
.c1:active {
  border: none;
  cursor: default;
  outline: none;
}

.c1::before {
  position: absolute;
  width: 100%;
  height: 100%;
  -webkit-animation: 3000ms ease-in-out dQLxgQ infinite;
  animation: 3000ms ease-in-out dQLxgQ infinite;
  background-color: #f3f3f3;
  border-radius: 3px;
  content: '';
}

<div>
  <div
    class="c0"
    height="400"
    type="box"
    width="400"
  >
    <div
      class="c1"
    />
  </div>
</div>
`;

exports[`Image should render with defined props 1`] = `
.c0 {
  display: block;
  height: 400px;
  width: 400px;
  position: relative;
  background-color: none;
}

.c1 > * {
  object-fit: cover;
  height: 100%;
  width: 100%;
  object-position: 50% 50%;
}

.c2 {
  opacity: 0;
}

<div>
  <div
    class="c0 c1"
    height="400px"
    width="400px"
  >
    <img
      class="c2"
      height="100%"
      opacity="0"
      src="https://piste-ui.s3.us-west-2.amazonaws.com/ski.jpg"
      width="100%"
    />
    
  </div>
</div>
`;

exports[`Image should render with preview 1`] = `
.c0 {
  display: block;
  height: 400px;
  width: 400px;
  position: relative;
  background-color: none;
}

.c1 > * {
  object-fit: cover;
  height: 100%;
  width: 100%;
  object-position: 50% 50%;
}

.c3 {
  position: absolute;
  left: 0px;
  top: 0px;
  -webkit-filter: blur(5px);
  filter: blur(5px);
  opacity: 1;
  -webkit-transition: opacity ease-in 200ms;
  transition: opacity ease-in 200ms;
  -webkit-clip-path: inset(0);
  clip-path: inset(0);
}

.c2 {
  opacity: 0;
}

<div>
  <div
    class="c0 c1"
    height="400px"
    width="400px"
  >
    <img
      class="c2"
      height="100%"
      opacity="0"
      src="https://piste-ui.s3.us-west-2.amazonaws.com/ski.jpg"
      width="100%"
    />
    <img
      class="c3"
      height="100%"
      opacity="1"
      src="https://piste-ui.s3.us-west-2.amazonaws.com/ski--thumbnail.jpg"
      width="100%"
    />
  </div>
</div>
`;
